### PR TITLE
Feature/frontend

### PR DESCRIPTION
**Pull request by Mat24**

### CODE REVIEW

File: apps/weather_service/lib/weather_service.ex [hunk 1]
```diff
@@ -12,8 +14,13 @@ defmodule WeatherService do
...
-  @spec get_weather(map) :: {:ok, map} | {:error, atom}
-  def get_weather(params) do
-    {:ok, %{id: 1}}
+  @spec get_weather(String.t, String.t) :: {:ok, String.t} | {:error, atom}
+  def get_weather(correlation_id, city) do
+    Logger.info("Getting Weather: #{correlation_id}")
+    with {:ok, weather} <- Weather.weather(%{city: city}, correlation_id) do
+      {:ok, weather}
+    else
+      err -> err
```
Comment (ever-cifuentes-globant) [inline apps/weather_service/lib/weather_service.ex]: The `else` statement could be removed here, at least you want to personalize the response error

File: apps/weather_service/lib/weather_service/utils.ex [hunk 2]
```diff
@@ -0,0 +1,21 @@
+defmodule Utils do
+  def kelvin_to_celcius(kelvin_grades), do: round(kelvin_grades - 273.15)
+
+  def normal_time(unix_time) do
+    with {:ok, date} <- DateTime.from_unix(unix_time),
+         {:ok, time} <- get_hours(date)
+    do
+      time
+    else
+      err -> err
```
Comment (ever-cifuentes-globant) [inline apps/weather_service/lib/weather_service/utils.ex]: Same here, the `else` could be omitted.

File: apps/weather_service/lib/weather_service/utils.ex [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    end
+
+  end
+
+  defp get_hours(date_time) do
+    hours = date_time.hour
```
Comment (ever-cifuentes-globant) [inline apps/weather_service/lib/weather_service/utils.ex]: `hours` and `minutes` could be removed and put them directly in the interpolation.

File: apps/weather_service/lib/weather_service/weather.ex [hunk 4]
```diff
@@ -1,3 +1,78 @@
 defmodule Weather do
-  
+  import Logger
+  # Error handling when city is invalid or nil
+  def weather(params, correlation_id) do
+    Logger.info("Task ID: #{correlation_id} has been started "<>
+    "with: #{inspect params}")
```
Comment (ever-cifuentes-globant) [inline apps/weather_service/lib/weather_service/weather.ex]: align the string with the start of the string of the previous line, "Task..."

File: apps/weather_service/lib/weather_service/weather.ex [hunk 5]
```diff
@@ -1,3 +1,78 @@
...
+    with {:ok, validated_params} <-  WeatherRequest.validate_params(params),
+         {:ok, response} <- get_from_api(validated_params),
+         {:ok, weather_element} <- weather_data_cleaning(response),
+         {:ok, processed_weather} <- weather_data_processor(weather_element)
+    do
+      Logger.info("Task ID: #{correlation_id} has a update: #{inspect processed_weather}")
+      {:ok, processed_weather}
+    else
+      err -> IO.inspect err
```
Comment (ever-cifuentes-globant) [inline apps/weather_service/lib/weather_service/weather.ex]: the return error could be replaced to `{:error, :internal_server_error}`

File: apps/weather_service/lib/weather_service/weather.ex [hunk 6]
```diff
@@ -1,3 +1,78 @@
...
+    endpoint = "#{get_api_url()}weather?q=#{city},co#{get_api_key()}"
+    
+    case HTTPoison.get(endpoint) do
+      {:ok, response} -> 
+        case Poison.decode!(response.body) do
+          %{"cod" => _, "message" => _} -> {:error, :not_found}
+          response -> {:ok, response}
+        end
+      error -> 
+        {:error, error}
```
Comment (ever-cifuentes-globant) [inline apps/weather_service/lib/weather_service/weather.ex]: The return error tuple must be in the same line of the pattern matching when it's enough short.

```elixir
error -> {:error, error}
```

File: apps/weather_service/lib/weather_service/weather_element.ex [hunk 7]
```diff
@@ -0,0 +1,82 @@
+defmodule WeatherElement do
```
Comment (ever-cifuentes-globant) [inline apps/weather_service/lib/weather_service/weather_element.ex]: Check this module, I guess the most of this code is obsolete right now.

File: apps/weather_service/lib/weather_service/weather_request.ex [hunk 8]
```diff
@@ -0,0 +1,17 @@
+defmodule WeatherRequest do
+  defstruct city: nil
+
+  def validate_params(params) do
+    with {:ok, city} <- parse_city(params[:city]) do
+      {:ok, %WeatherRequest{city: city}}
+    else
+      err -> err
```
Comment (ever-cifuentes-globant) [inline apps/weather_service/lib/weather_service/weather_request.ex]: Same here with the `else`

File: apps/weather_service_web/spec/controllers/weather_controller_spec.exs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+      it "should return a valid temperature" do
+        :index 
+        |> response_get(request())
+        |> get_body()
+        |> Map.get("temperature")
+        |> String.split()
+        |> Enum.at(0) 
+        |> String.to_integer()
+        |> is_integer()
+        |> IO.inspect()
```
Comment (ever-cifuentes-globant) [inline apps/weather_service_web/spec/controllers/weather_controller_spec.exs]: Do not forget to remove the `IO.inspect` printings

File: apps/weather_service_web/spec/support/factory.ex [hunk 10]
```diff
@@ -0,0 +1,17 @@
+defmodule WeatherServiceWeb.Factory do
+  use ExMachina
+  def weather_factory do
+    %{
+      location_name: Enum.random(["Bogota CO", "Cali CO", "Medellin CO"]),
+      temperature: "#{Enum.random(5..40)} °C",
+      wind: "",
```
Comment (ever-cifuentes-globant) [inline apps/weather_service_web/spec/support/factory.ex]: Complete the missing factory weather fields

**Review by ever-cifuentes-globant — CHANGES_REQUESTED** [review]

**Review by ever-cifuentes-globant — APPROVED** [review]

**Comment by ricardo-mender-globant** [comment]
Lgtm, nice work!

**Pull request merged**